### ISSUE TITLE
Investigate optimizer pass to make unaccessed named fields anonymous

### ISSUE THREAD

**Issue by bbannier** [zeek/spicy#1409]
In #1398 we discussed an optimizer pass to make unaccessed named field anonymous.  We should have a look at how hard this would be to add, maybe even without adding control flow analysis.

**Comment by rsmmr** [comment]
> So in this case, we can document a way to "use" the fields and just say if you want them, use them (like a `_ = self.this_is_unused` kind of thing).

The nice thing about an option is that it would make semantics very clear, including for further optimizations that we may be adding over time: either people can rely on the generated C++ code or they can't. Documenting ways to get around certain individual optimizations feels brittle in contrast, and also hard to explain.

>  Part of me wonders if we add another flag, then the "less aggressive optimizations" flag would just be unused in practice.

Yeah, I see that risk, although it's probably pretty well testable at least to ensure we don't break it: when a new optimization comes around that can change the C++ interface, it should be easy to add tests for both cases.

There is a general question if we want to keep the "use-the-C++-code-directly" use case going forward; there's clearly not much demand for it so far. But I'm not quite ready to let that go.

**Comment by evantypanski** [comment]
> Did you already have some code for this?

Yes, I can't seem to find it so I guess I never made a branch even locally.

I don't remember the exact implementation, but it just collected all uses of fields then marked any unused fields as transient in the Spicy side, just to see what the potential side effects would be
